### ISSUE TITLE
no-unused-prop-types Rule does not recognise component that returns an array

### ISSUE THREAD

**Issue by leonaves** [yannickcr/eslint-plugin-react#1682]
Example:

```js
import React from 'react';
import PropTypes from 'prop-types';

const TestComponent = () => (
  ['one', 'two', 'three'].map(text => <div>{ text }</div>)
);

TestComponent.propTypes = {
  customClass: PropTypes.string,
};

export default TestComponent;
```

React 16 allows for arrays to be returned from components, but in the above example, the `no-unused-prop-types` seems to not recognise that `customClass` is unused. It does detect it when the component is changed to:

```js
const TestComponent = () => {
  const componentArr = ['one', 'two', 'three'].map(text => <div>{ text }</div>);
  return <div>{ componentArr }</div>;
};
```

**Comment by ljharb** [comment]
Our component detection likely requires a component to return JSX.

It becomes very very difficult to detect if an SFC that returns no JSX is in fact a component. A temporary workaround should be to add a comment above the SFC that says `/* @extends React.Component */`.

**Comment by nodkrot** [comment]
This similarly applies to this rule as well https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/prop-types.md

**Comment by ljharb** [comment]
Note that you can also return a React.Fragment or `<>`.

**Comment by nodkrot** [comment]
That could work as a wrapper 👍

**Comment by leonaves** [comment]
Forgot about this, but re-reading, why do you need to detect that the function is a component? If the prop is unused in the function body then the rule should error, it shouldn't matter if it returns JSX or not.

**Comment by ljharb** [comment]
@leonaves if it's not a component, then it doesn't have "props" - it has arguments, and it can't have propTypes.
